### PR TITLE
KEP-276 Refactor CRUD Library

### PR DESCRIPTION
**Pull request by rnistuk**

### CODE REVIEW

File: crud/crud.cpp [hunk 1]
```diff
@@ -398,21 +398,14 @@ crud::set_leader_info(bzn::message& msg)
 
 
 bool
-crud::validate_create(const bzn::message &msg)
+crud::validate_create_or_update(const bzn::message &msg)
```
Comment (ebruck) [inline crud/crud.cpp]: please move reference to the type throughout the code

File: crud/crud.cpp [hunk 2]
```diff
@@ -61,7 +52,7 @@ crud::start()
                 {
                     LOG(debug) << " commit: " << msg.toStyledString();
 
-                    if (auto search = self->handlers.find(msg["cmd"].asString());search != self->handlers.end())
+                    if (auto search = self->commit_handlers.find(msg["cmd"].asString());search != self->commit_handlers.end())
```
Comment (ebruck) [inline crud/crud.cpp]: I prefer:  
if (auto it = ... 

"it" is more common to indicate an iterator.

File: crud/crud.cpp [hunk 3]
```diff
@@ -83,8 +74,15 @@ crud::set_leader_info(bzn::message& msg)
...
+    return false;
+}
+
+
+
 void
-crud::handle_create(const bzn::message& msg)
+crud::commit_create(const bzn::message &msg)
 {
     storage_base::result result = this->storage->create(msg["db-uuid"].asString(), msg["data"]["key"].asString(), msg["data"]["value"].asString());
```
Comment (ebruck) [inline crud/crud.cpp]: auto result = ...

File: crud/crud.cpp [hunk 4]
```diff
@@ -308,7 +306,8 @@ crud::do_follower_tasks(const bzn::message& msg)
     // TODO: replace this if else tree with the strategy pattern.
     if(cmd==bzn::MSG_CMD_READ)
     {
-        this->handle_read(msg, response);
+        this->command_handlers[bzn::MSG_CMD_READ](msg, response);
+        //this->handle_read(msg, response);
```
Comment (ebruck) [inline crud/crud.cpp]: remove dead code

File: crud/crud.cpp [hunk 5]
```diff
@@ -58,11 +58,31 @@ crud::start()
...
         });
 }
 
+void
+crud::do_raft_task_routing(const bzn::message& msg, bzn::message& response)
+{
+    switch(raft->get_state())
+    {
+        case bzn::raft_state::candidate:
+            response = do_candidate_tasks(msg);
```
Comment (ebruck) [inline crud/crud.cpp]: Missing "this":
ie.
this->do_candidate_tasks();

File: crud/crud.cpp [hunk 6]
```diff
@@ -209,6 +230,19 @@ crud::commit_delete(const bzn::message &msg)
     }
 }
 
+void
+crud::handle_delete(const bzn::message &msg, bzn::message &response)
+{
+    if (this->storage->has(msg["db-uuid"].asString(), msg["data"]["key"].asString()))
+    {
+        this->raft->append_log(msg);
+    }
```
Comment (ebruck) [inline crud/crud.cpp]: You could to an early return here and drop the else.

File: crud/crud.cpp [hunk 7]
```diff
@@ -243,45 +277,28 @@ crud::handle_size(const bzn::message& msg, bzn::message& response)
...
+    if( !(msg.isMember("bzn-api") && msg["bzn-api"]=="crud"))
     {
-        this->do_raft_task_routing(msg, *response);
+        (*response)["error"] = bzn::MSG_INVALID_CRUD_COMMAND;
     }
     else
     {
-        (*response)["request-id"] = msg["request-id"];
-        (*response)["error"] = bzn::MSG_INVALID_CRUD_COMMAND;
+        if( auto search = accepted_crud_commands.find(msg["cmd"].asString()); search != accepted_crud_commands.end() )
```
Comment (ebruck) [inline crud/crud.cpp]: search is not used... do:

if (this->accept_crud_commands.find() != this->accept_crud_commands.end())
...

File: crud/crud.cpp [hunk 8]
```diff
@@ -65,24 +64,26 @@ crud::start()
...
+    response["error"] = error;
+}
+
+
+void
+crud::do_raft_task_routing(const bzn::message& request, bzn::message& response)
 {
-    switch(raft->get_state())
+    // TODO: This module can be simplified by removing the need for this method. We can/do check state *inside* of the command handlers.
+    auto search = this->route_handlers.find(raft->get_state());
```
Comment (ebruck) [inline crud/crud.cpp]: prefer auto it = this->... find();

File: crud/crud.cpp [hunk 9]
```diff
@@ -191,18 +193,13 @@ crud::handle_read(const bzn::message& msg, bzn::message& response)
     }
 }
 
+
 void
 crud::commit_update(const bzn::message &msg)
 {
-    LOG(debug) << msg.toStyledString();
-
     storage_base::result result = this->storage->update(msg["db-uuid"].asString(), msg["data"]["key"].asString(), msg["data"]["value"].asString());
```
Comment (ebruck) [inline crud/crud.cpp]: auto result = ...

File: crud/test/crud_tests.cpp [hunk 10]
```diff
@@ -138,6 +138,9 @@ TEST_F(crud_test, test_that_follower_not_knowing_leader_fails_to_create)
     EXPECT_CALL(*this->mock_session, send_message(_,_)).WillOnce(Invoke(
         [&](auto& msg, auto)
         {
+//            LOG(debug) << "msg: " << (*msg).toStyledString();
```
Comment (ebruck) [inline crud/test/crud_tests.cpp]: lets remove dead code

File: crud/crud.cpp [hunk 11]
```diff
@@ -118,77 +121,53 @@ crud::commit_create(const bzn::message &msg)
...
-    return msg.isMember("db-uuid") && msg.isMember("data") && msg["data"].isMember("key") && msg["data"].isMember("value");
-}
-
-
 void
-crud::handle_create(const bzn::message &msg, bzn::message &response)
+crud::handle_read(const bzn::message& request, bzn::message& response)
 {
-    if(!this->validate_create(msg))
+    if(!validate_read_xor_delete(request))
```
Comment (ebruck) [inline crud/crud.cpp]: xor?

File: crud/crud.cpp [hunk 12]
```diff
@@ -303,37 +289,25 @@ crud::handle_ws_crud_messages(const bzn::message& msg, std::shared_ptr<bzn::sess
...
 
-    // TODO: replace this if else tree with the strategy pattern.
     if(cmd==bzn::MSG_CMD_READ)
     {
         this->command_handlers[bzn::MSG_CMD_READ](request, response);
-        //this->handle_read(msg, response);
     }
-    else if(cmd==bzn::MSG_CMD_KEYS)
+    // TODO: Move utilities into their own module.
+    else if (auto search = this->utility_handlers.find(cmd); search != this->utility_handlers.end())
```
Comment (ebruck) [inline crud/crud.cpp]: prefer: auto it = ...

File: crud/crud.cpp [hunk 13]
```diff
@@ -346,34 +320,22 @@ crud::do_follower_tasks(const bzn::message& request, bzn::message& response)
...
     if(request.isMember("cmd"))
     {
-
-
-
-        const auto command_handler = this->command_handlers.find(request["cmd"].asString());
-
-        if (command_handler == this->command_handlers.end())
+        std::string command = request["cmd"].asString();
+        if(auto ch = this->command_handlers.find(command); ch != this->command_handlers.end())
```
Comment (ebruck) [inline crud/crud.cpp]: you can use "auto it" as ch is local only to the if.

File: crud/crud.cpp [hunk 14]
```diff
@@ -346,34 +320,22 @@ crud::do_follower_tasks(const bzn::message& request, bzn::message& response)
...
         {
-            this->raft->append_log(request);
+            ch->second(request, response);
+        }
+        // TODO: Move utilities into their own module
+        else if(auto uh = this->utility_handlers.find(command); uh != this->utility_handlers.end())
```
Comment (ebruck) [inline crud/crud.cpp]: you can use "auto it" as uh is local only to the if.

File: crud/crud.cpp (hunk 1 above)
Comment (rnistuk) [reply crud/crud.cpp]: Done in crud.cpp, crud.h and crud_test.cpp

File: crud/crud.cpp (hunk 3 above)
Comment (rnistuk) [reply crud/crud.cpp]: Went through crud.cpp and crud_test.cpp and changed the local variable declarations to auto.

File: crud/crud.cpp (hunk 5 above)
Comment (rnistuk) [reply crud/crud.cpp]: I think you may have been working with an older version, or maybe I didn't push correctly. I removed the switch statements and replaced them with lookup tables.
This is no longer an issue...

File: crud/test/crud_tests.cpp [hunk 15]
```diff
@@ -648,8 +639,8 @@ TEST_F(crud_test, test_that_a_leader_can_delete_an_existing_record)
     //   do concensus via RAFT
     //   when RAFT calls commit handler perform delete
 
-    const std::string key{"key0"};
-    bzn::message request = generate_delete_request(user_uuid, key);
+    const auto key{"key0"};
```
Comment (ebruck) [inline crud/test/crud_tests.cpp]: What type were you expecting here?

**Pull request merged**